src/coreComponents/constitutive/fluid/multifluid/constant/InvariantImmiscibleFluid.cpp

```diff
@@ -0,0 +1,91 @@
+#include "InvariantImmiscibleFluid.hpp"
+#include "common/format/StringUtilities.hpp"
+#include "constitutive/ConstitutiveBase.hpp"
+
+namespace geos
+{
+namespace constitutive
+{
+
+InvariantImmiscibleFluid::InvariantImmiscibleFluid( string const & name, Group * parent )
+  : MultiFluidBase( name, parent )
+{
+  // Override input flags for mandatory options
+
+  registerWrapper( viewKeyStruct::componentNamesString(), &m_componentNames )
+    .setInputFlag( dataRepository::InputFlags::REQUIRED )
+    .setDescription( "List of fluid components (e.g. CH4, H2O, C5H12)" );
+
+  registerWrapper( viewKeyStruct::phaseNamesString(), &m_phaseNames )
+    .setInputFlag( dataRepository::InputFlags::REQUIRED )
+    .setDescription( "List of fluid phases (e.g. gas, water, oil)" );
+
+  registerWrapper( viewKeyStruct::componentMolarWeightString(), &m_componentMolarWeight )
+    .setInputFlag( dataRepository::InputFlags::REQUIRED )
+    .setDescription( "Molar weights of components" );
+
+  // Densities: constant phase densities
+  registerWrapper( "densities", &m_densities )
+    .setInputFlag( dataRepository::InputFlags::REQUIRED )
+    .setDescription( "Constant phase densities" );
+
+  // Viscosities: constant phase viscosities
+  registerWrapper( "viscosities", &m_viscosities )
+    .setInputFlag( dataRepository::InputFlags::REQUIRED )
+    .setDescription( "Constant phase viscosities" );
+}
+
+integer InvariantImmiscibleFluid::getWaterPhaseIndex() const
+{
+  // find index of water phase in user ordering
+  for( size_t i=0; i< m_phaseNames.size(); ++i )
+  {
+    if( stringutilities::toLower( m_phaseNames[i] ) == "water" )
+      return static_cast< integer >(i);
+  }
+  return -1;
+}
+
+void InvariantImmiscibleFluid::postInputInitialization()
+{
+  // check base inputs
+  MultiFluidBase::postInputInitialization();
+
+  integer numPhase = numFluidPhases();
```

Comment:
I think we have an assumption that number of components equals number of phases. So we need to check `numFluidPhases()` against `numFluidComponents()`.

Reply:
The assumption is now checked at the specified location, and the test has been updated accordingly.